File: 0148-sort-list/0148-sort-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 function getMiddle(head){
        let slow = head;
        let fast = head;
        while(fast!= null && fast.next!=null && fast.next.next!=null){
            slow= slow.next;
            fast = fast.next.next
        }
        return slow
    }
    var mergeTwoLists = function (list1, list2) {
    if (!list1) {
        return list2
    }
    if (!list2) {
        return list1
    }
    if (list1.val < list2.val) {
        list1.next = mergeTwoLists(list1.next, list2)
        return list1;
    }
    else {
        list2.next = mergeTwoLists(list2.next, list1);
        return list2;
    }


};
var sortList = function(head) {
    
    if (head==null || head.next==null){
        return head;
    }

    var middle = getMiddle(head);

    var headofsecondpart = middle.next;

    middle.next = null;

    var l1= head;
    var l2= headofsecondpart;

    return mergeTwoLists(sortList(l1),sortList(l2) )

    
};

File: 0148-sort-list/0148-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* l1, ListNode* l2) {
        if (l1 == NULL && l2 == NULL)
            return NULL;
        if (l1 == NULL)
            return l2;
        if (l2 == NULL)
            return l1;
        if (l1->val <= l2->val) {
            l1->next = merge(l1->next, l2);
            return l1;
        } else {
            l2->next = merge(l2->next, l1);
            return l2;
        }
    }
    ListNode* getMiddle(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    ListNode* sortList(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode* middle = getMiddle(head);
        ListNode* middlenext = middle->next;
        middle->next = NULL;
        ListNode* l1 = sortList(head);
        ListNode* l2 = sortList(middlenext);
        return merge(l1, l2);
    }
};